File: Practica 12/conVolParalelo.java
/**
*	Calculo de convoluciones de matrices de forma secuencial.
*	
*	@author José Miguel Aragón Jurado
*	@version 1.0.
*/

import java.util.*;
import java.util.concurrent.*;

public class conVolParalelo implements Runnable{
	private static int[][] m;
	private static int[][] r;
	private static int[][] k;
	private int inicio;
	private int fin;

	private static int[][] k_enfoque={{0,-1,0},
	{-1,5,-1},
	{0,-1,0}};

	private static int[][] k_desenfoque={{1,1,1},
	{1,1,1},
	{1,1,1}};
	private static int[][] k_realce={{0,0,1},
	{-1,1,0},
	{0,0,0}};
	private static int[][] k_detec={{0,1,0},
	{1,-4,1},
	{0,1,0}};
	private static int[][] k_sobel={{-1,0,1},
	{-2,0,2},
	{-1,0,1}};
	private static int[][] k_sharpen={{1,-2,1},
	{-2,5,-2},
	{1,-2,1}};

	public conVolParalelo(int inicio,int fin){
		this.inicio=inicio;
		this.fin=fin;
	}

	public void run(){
		for (int i=inicio;i<fin;i++){
			for(int j=0;j<m[0].length;j++){
				nextState(i,j);
			}		
		}
	}

	private static void rellenarMatriz(){
		Random r = new Random();
		for(int i=0;i<m.length;i++){
			for(int j=0;j<m[0].length;j++){
				m[i][j]=r.nextInt(41)-20;
			}
		}
	}

	private static void mostrarMatriz(int[][] m){
		for(int i=0;i<m.length;i++){
			for(int j=0;j<m[0].length;j++){
				System.out.print(m[i][j]+",");
			}
			System.out.println();
		}
		System.out.println();
	}

	private static void convolucion(int[][]kernel,int nh){
		k=kernel;
		long tiempo = System.nanoTime();
		ExecutorService ex = Executors.newFixedThreadPool(nh);
		int bloque = m.length/nh;
		for(int i=0;i<nh;i++){
			if(i==nh-1){
				ex.execute(new conVolParalelo(i*bloque,m.length));
			}
			else{
				ex.execute(new conVolParalelo(i*bloque,(i+1*bloque)));
			}
		}
		ex.shutdown();
		while(!ex.isTerminated());
		tiempo=System.nanoTime()-tiempo;
		System.out.println("t:"+tiempo/1e09);
	}


	private static void nextState(int a,int b){
		int z=0;
		for(int i=a-1;i<=a+1;i++){
			int l=0;
			for(int j=b-1;j<=b+1;j++){ 
				try{
					r[a][b]=r[a][b]+m[i][j]*k[z][l];
				}catch(IndexOutOfBoundsException e){}
				finally{
					l++;
				}
			}
			z++;
		}
	}
	public static void main(String[] args) {
		m=new int[10000][10000];
		r=new int[10000][10000];
		k=new int[3][3];
		rellenarMatriz();
		//mostrarMatriz(m);
		Scanner sc = new Scanner(System.in);
		System.out.println("Elige que kernel desea utilizar:");
		System.out.println("1. Enfoque.");
		System.out.println("2. Desenfoque.");
		System.out.println("3. Realce de bordes.");
		System.out.println("4. Deteccion de bordes.");
		System.out.println("5. Sobel.");
		System.out.println("6. Sharpen.");
		int elec=sc.nextInt();
		switch(elec){
			case 1:
			convolucion(k_enfoque,Integer.parseInt(args[0]));
			//mostrarMatriz(r);
			break;
			case 2:
			convolucion(k_desenfoque,Integer.parseInt(args[0]));
			//mostrarMatriz(r);
			break;
			case 3:
			convolucion(k_realce,Integer.parseInt(args[0]));
			//mostrarMatriz(r);
			break;
			case 4:
			convolucion(k_detec,Integer.parseInt(args[0]));
			//mostrarMatriz(r);
			break;
			case 5:
			convolucion(k_sobel,Integer.parseInt(args[0]));
			//mostrarMatriz(r);
			break;
			case 6:
			convolucion(k_sharpen,Integer.parseInt(args[0]));
			//mostrarMatriz(r);
			break;

		}

	}
}

File: Practica 12/conVolParalelo.cpp
/*Convolución de matrices paralelo en c++
	Autor: José Miguel Aragón Jurado
*/
#define N 10000
#include <iostream>
#include <random>
#include <cstdlib>
#include <ctime>
#include <thread>
using namespace std;
int m[N][N];
int r[N][N];
int k1[3][3]={{0,-1,0},
{-1,5,-1},
{0,-1,0}};
int k2[3][3]={{1,1,1},
{1,1,1},
{1,1,1}};
int k3[3][3]={{0,0,1},
{-1,1,0},
{0,0,0}};
int k4[3][3]={{0,1,0},
{1,-4,1},
{0,1,0}};

int k5[3][3]={{-1,0,1},
{-2,0,2},
{-1,0,1}};
int k6[3][3]={{1,-2,1},
{-2,5,-2},
{1,-2,1}};


void nextState(int,int,int[][3]);
void rellenarMatriz();
void mostrarMatriz();
void convolucion(int[][3],int); 
void run(int[][3] ,int ,int);


int main(int argc,char* args[]){
	if(argc==1){
		cout<<"Introduce el numero de hebras"<<endl;
	}
	else{

		rellenarMatriz();
		cout<<"Introduce que tecnica de convolucion desea aplicar:"<<endl;
		cout<<"1. Enfoque."<<endl;
		cout<<"2. Desenfoque."<<endl;
		cout<<"3. Realizar bordes."<<endl;
		cout<<"4. Detección de bordes."<<endl;
		cout<<"5. Sobel."<<endl;
		cout<<"6. Sharpen."<<endl;
		int elec;
		cin>>elec;
		switch(elec){
			case 1:{
				convolucion(k1,atoi(args[1]));
				break;
			}
			case 2:{
				convolucion(k2,atoi(args[1]));
				break;
			}
			case 3:
			{
				convolucion(k3,atoi(args[1]));
				break;
			}
			case 4:{
				convolucion(k4,atoi(args[1]));
				break;
			}
			case 5:{
				convolucion(k5,atoi(args[1]));
				break;
			}
			case 6:{
				convolucion(k6,atoi(args[1]));
				break;
			}
		}
	}

	return 0;
}

void rellenarMatriz(){
	uniform_real_distribution<> random(-20,20);
	default_random_engine seed(time(0));
	for(int i=0;i<N;i++){
		for(int j=0;j<N;j++){
			m[i][j]=round(random(seed));
			r[i][j]=0;
		}
	}
}
void mostrarMatriz(){
	for(int i=0;i<N;i++){
		for(int j=0;j<N;j++){
			cout<<m[i][j]<<",";
		}
		cout<<endl;
	}
	cout<<endl;
}

void convolucion(int k[][3],int nh){
	chrono::time_point<chrono::system_clock> start,end;
	start=chrono::system_clock::now();
	thread hilos[nh];
	int bloque=(int)N/nh;
	for(int i=0;i<nh;i++){
		if(i==nh-1){
			hilos[i]=thread(run,ref(k),i*bloque,N);
		}
		else{
			hilos[i]=thread(run,ref(k),i*bloque,(i+1)*bloque);
		}
	}
	for(int i=0;i<nh;i++){
		hilos[i].join();
	}
	end=chrono::system_clock::now();
	chrono::duration<double> tiempo = end-start;
	cout<<"t= "<<tiempo.count()<<endl;


}

void run(int k[][3],int inicio,int fin){
	for(int i=inicio;i<fin;i++){
		for(int j=0;j<N;j++){
			nextState(i,j,k);
		}
	}
}

void nextState(int a,int b,int k[][3]){
	int z=0;
	for(int i=a-1;i<=a+1;i++){
		int l=0;
		for(int j=b-1;j<=b+1;j++){
			if(i>=N || j>=N|| j<0 || i<0);
			else{
				r[a][b]=r[a][b]+k[z][l]*m[i][j];
			}
			l++;
		}
		z++;
	}
}